docs/design/2020-04-17-sql-sequence.md

```diff
@@ -0,0 +1,118 @@
+
+create sequence statement:
+
+```
+CREATE SEQUENCE [IF NOT EXISTS] sequence_name
+[ INCREMENT [ BY | = ] increment ]
+[ MINVALUE [=] minvalue | NO MINVALUE | NOMINVALUE ]
+[ MAXVALUE [=] maxvalue | NO MAXVALUE | NOMAXVALUE ]
+[ START [ WITH | = ] start ] 
+[ CACHE [=] cache | NOCACHE | NO CACHE ]
+[ CYCLE | NOCYCLE | NO CYCLE]
+```
+ 
+show create sequence statement:
+
+```
+SHOW CREATE SEQUENCE sequence_name
+```
+
+drop sequence sequence statement:
+
+```
+DROP SEQUENCE sequence_name_list
+```
+
+get the next value of sequence:
+
+```
+NEXT VALUE FOR sequence_name / NEXTVAL(sequence_name)
+```
+
+get the previous value of sequence:
+
+```
+LASTVAL(sequence_name)
+```
+
+move the current position in sequence number axis
+
+```
+SETVAL(sequence_name, num)
+```
+
+## Rationale
+
+- Sequence metadata
+
+Sequence is a table-level object, sharing the same namespace with the base table and view objects, whose meta are recorded in `ast.tableInfo`. But actually sequence definition do not have columns in it's meta, which means you can't operate the sequence value by `DML` interface. The only way you can do it is by `setval` function, after all, the sequence value is essentially a `KV` pair with some constraints.
+
+- Sequence values
+
+Each sequence value will be stored in its own range, with the key format below:
+
+```
+DB:<dbID> SID:<sequenceID> ——> int64 
+```
+
+Reads and writes to the sequence value (via the functions `nextval`, `setval`), will be implemented by direct calls to the `KV` layer's `Get`, `Inc`, and `Set` functions. Since sequence object store the value at the range of its own, it will cause hotpoint at sequence meta under insertion intensive cases.
+
+- Sequence Transaction
+
+Storing sequence values in their own ranges means that inserts to tables which use sequences will always touch two ranges. However, since the sequence update takes place outside of the SQL transaction, this should not trigger the `2PC` commit protocol. 
+
+Sequence computation is under the expression framework, evaluating one value at a time. The values fetched by these functions won't be rolled back when the outer SQL transaction is aborted.
+
+- Sequence Binlog
+
+Distinguished from auto-increment's rebase functionality, columns with sequence as default won't rebase it's value automatically when the inserted value was specified explicitly. Besides, sequence can also be used directly as the inserted expression value. Thus, it's necessary to sync the sequence value down to the upstream cluster.
+
+Sequence binlog is triggered by both `nextval` and `setval` functions when the update reaches the storage directly rather than in the cache.
+
+Sequence binlog is not a sort of `DML` binlog, but mocked as a kind of `DDL` `job` by specifying the job id equal to -1 specially. Therefore, the drainer won't take it for granted that it's a real ddl job and it should be occurred in ddl history queue. 
```

Comment:
What do you mean by `it should be occurred in ddl history queue`? It should not appear in ddl history queue.

Reply:
shouldn't.
take it for granted A & B